File: server/internal/broker/oidc_test.go
package broker

import (
	"context"
	"net/url"
	"testing"
)

// TestBuildAuthorizeURL mirrors oidc.test.ts buildAuthorizeUrl: response_type,
// PKCE S256, state, scope, and (when set) audience+resource.
func TestBuildAuthorizeURL(t *testing.T) {
	oc := newOIDCClientRaw(oidcConfig{
		AuthorizeURL: "https://idp.example/authorize",
		ClientID:     "lumen-web",
	})
	got, err := oc.buildAuthorizeURL(authorizeParams{
		CodeChallenge: "chal",
		State:         "st8",
		RedirectURI:   "https://acct.example.com/desktop/callback",
		Scope:         desktopScope,
		Audience:      "lumen-api",
	})
	if err != nil {
		t.Fatalf("build: %v", err)
	}
	u, _ := url.Parse(got)
	q := u.Query()
	checks := map[string]string{
		"response_type":         "code",
		"client_id":             "lumen-web",
		"redirect_uri":          "https://acct.example.com/desktop/callback",
		"scope":                 desktopScope,
		"state":                 "st8",
		"code_challenge":        "chal",
		"code_challenge_method": "S256",
		"audience":              "lumen-api",
		"resource":              "lumen-api",
	}
	for k, want := range checks {
		if got := q.Get(k); got != want {
			t.Errorf("query %s = %q, want %q", k, got, want)
		}
	}
}

// TestBuildAuthorizeURLNoAudience verifies audience/resource are omitted for the
// account-center flow (no aud=lumen-api).
func TestBuildAuthorizeURLNoAudience(t *testing.T) {
	oc := newOIDCClientRaw(oidcConfig{AuthorizeURL: "https://idp.example/authorize", ClientID: "web"})
	got, _ := oc.buildAuthorizeURL(authorizeParams{
		CodeChallenge: "c", State: "s", RedirectURI: "https://acct/cb", Scope: webScope,
	})
	u, _ := url.Parse(got)
	if u.Query().Has("audience") || u.Query().Has("resource") {
		t.Error("account-center authorize URL must not carry audience/resource")
	}
}

// TestExchangeAuthCode mirrors oidc.ts exchangeAuthCode against a stub IdP,
// verifying the correct grant + form fields and a decoded token.
func TestExchangeAuthCode(t *testing.T) {
	idp := newIDPStub(t)
	idp.tokenResp = map[string]any{
		"access_token":  fakeJWT(map[string]any{"sub": "u-1"}),
		"refresh_token": "refresh-abc",
		"id_token":      fakeJWT(map[string]any{"sub": "u-1", "name": "Alice"}),
		"expires_in":    float64(3600),
	}
	oc := idp.newTestOIDC()
	tok := oc.exchangeAuthCode(context.Background(), "auth-code-1", "verifier-1", "https://acct/cb")
	if tok == nil {
		t.Fatal("token is nil")
	}
	if tok.AccessToken == "" || tok.RefreshToken != "refresh-abc" {
		t.Errorf("token = %+v", tok)
	}
	// Form assertions: correct grant, code, verifier, client credentials.
	f := idp.lastTokenForm
	if f.Get("grant_type") != "authorization_code" {
		t.Errorf("grant_type = %q", f.Get("grant_type"))
	}
	if f.Get("code") != "auth-code-1" || f.Get("code_verifier") != "verifier-1" {
		t.Errorf("code/verifier = %q/%q", f.Get("code"), f.Get("code_verifier"))
	}
	if f.Get("client_id") != "lumen-web" || f.Get("client_secret") != "s3cr3t" {
		t.Errorf("client creds = %q/%q", f.Get("client_id"), f.Get("client_secret"))
	}
}

// TestExchangeAuthCodeRejected verifies a non-2xx IdP response yields nil.
func TestExchangeAuthCodeRejected(t *testing.T) {
	idp := newIDPStub(t)
	idp.tokenStatus = 400
	idp.tokenResp = map[string]any{"error": "invalid_grant"}
	oc := idp.newTestOIDC()
	if tok := oc.exchangeAuthCode(context.Background(), "bad", "v", "https://acct/cb"); tok != nil {
		t.Errorf("rejected exchange should be nil, got %+v", tok)
	}
}

// TestSubjectFromAndProfile mirrors oidc.ts subjectFrom/profileFromJwt.
func TestSubjectFromAndProfile(t *testing.T) {
	idToken := fakeJWT(map[string]any{"sub": "id-sub", "name": "Alice", "picture": "https://img/a.png"})
	accessToken := fakeJWT(map[string]any{"sub": "at-sub"})

	// id_token preferred for sub.
	if sub := subjectFrom(idToken, accessToken); sub != "id-sub" {
		t.Errorf("subjectFrom = %q, want id-sub", sub)
	}
	// falls back to access_token when id_token lacks sub.
	noSubID := fakeJWT(map[string]any{"name": "x"})
	if sub := subjectFrom(noSubID, accessToken); sub != "at-sub" {
		t.Errorf("subjectFrom fallback = %q, want at-sub", sub)
	}
	// profile from claims.
	p := profileFromJWT(idToken, accessToken)
	if p.DisplayName != "Alice" || p.AvatarURL != "https://img/a.png" {
		t.Errorf("profileFromJWT = %+v", p)
	}
}

// TestFetchProfileUserinfoFallback mirrors callback.test.ts: when the JWT lacks
// profile claims, /userinfo supplies them.
func TestFetchProfileUserinfoFallback(t *testing.T) {
	idp := newIDPStub(t)
	idp.infoResp = map[string]any{"sub": "u-2", "preferred_username": "bob", "picture": "https://img/b.png"}
	oc := idp.newTestOIDC()

	accessToken := fakeJWT(map[string]any{"sub": "u-2"})
	idToken := fakeJWT(map[string]any{"sub": "u-2"})
	p := oc.fetchProfile(context.Background(), accessToken, idToken)
	if p.DisplayName != "bob" || p.AvatarURL != "https://img/b.png" {
		t.Errorf("fetchProfile fallback = %+v", p)
	}
}

// TestProfileFromClaimsPrecedence mirrors oidc.ts profileFromClaims: name →
// preferred_username → nickname.
func TestProfileFromClaimsPrecedence(t *testing.T) {
	if p := profileFromClaims(map[string]any{"preferred_username": "pu", "nickname": "nk"}); p.DisplayName != "pu" {
		t.Errorf("display_name = %q, want pu", p.DisplayName)
	}
	if p := profileFromClaims(map[string]any{"nickname": "nk"}); p.DisplayName != "nk" {
		t.Errorf("display_name = %q, want nk", p.DisplayName)
	}
	if p := profileFromClaims(map[string]any{"name": "n", "preferred_username": "pu"}); p.DisplayName != "n" {
		t.Errorf("display_name = %q, want n", p.DisplayName)
	}
}

File: website/test/_lib/oidc.test.ts
import { describe, it, expect, afterEach } from "vitest";
import {
  buildAuthorizeUrl,
  subjectFrom,
  profileFromClaims,
  profileFromJwt,
  exchangeAuthCode,
  refreshWithIdp,
} from "../../functions/_lib/oidc";
import { fakeJwt, stubFetch, idpTokenRoute, makeEnv } from "./testutil";

const testEnv = makeEnv();

let restoreFetch: (() => void) | null = null;
afterEach(() => {
  restoreFetch?.();
  restoreFetch = null;
});

describe("buildAuthorizeUrl", () => {
  it("includes all required OAuth2 + PKCE params", () => {
    const url = new URL(
      buildAuthorizeUrl(testEnv, {
        codeChallenge: "challenge-value",
        state: "state-value",
        redirectUri: "https://test.example/desktop/callback",
        scope: "openid profile email offline_access",
        audience: "lumen-api",
      }),
    );
    expect(url.searchParams.get("response_type")).toBe("code");
    expect(url.searchParams.get("client_id")).toBe(testEnv.OIDC_CLIENT_ID);
    expect(url.searchParams.get("redirect_uri")).toBe("https://test.example/desktop/callback");
    expect(url.searchParams.get("scope")).toBe("openid profile email offline_access");
    expect(url.searchParams.get("state")).toBe("state-value");
    expect(url.searchParams.get("code_challenge")).toBe("challenge-value");
    expect(url.searchParams.get("code_challenge_method")).toBe("S256");
    expect(url.searchParams.get("audience")).toBe("lumen-api");
    expect(url.searchParams.get("resource")).toBe("lumen-api");
  });

  it("omits audience/resource when not provided (account center)", () => {
    const url = new URL(
      buildAuthorizeUrl(testEnv, {
        codeChallenge: "c",
        state: "s",
        redirectUri: testEnv.OIDC_WEB_REDIRECT_URI,
        scope: "openid profile email",
      }),
    );
    expect(url.searchParams.get("audience")).toBeNull();
    expect(url.searchParams.get("resource")).toBeNull();
  });
});

describe("subjectFrom", () => {
  it("prefers id_token sub, falls back to access_token", () => {
    expect(subjectFrom(fakeJwt({ sub: "from-id" }), fakeJwt({ sub: "from-access" }))).toBe("from-id");
    expect(subjectFrom(undefined, fakeJwt({ sub: "from-access" }))).toBe("from-access");
  });
  it("returns empty string when no sub present", () => {
    expect(subjectFrom(undefined, undefined)).toBe("");
    expect(subjectFrom("not-a-jwt")).toBe("");
  });
});

describe("profileFromClaims / profileFromJwt", () => {
  it("maps name → display_name and picture → avatar_url", () => {
    expect(profileFromClaims({ name: "Alice", picture: "https://img/a" })).toEqual({
      display_name: "Alice",
      avatar_url: "https://img/a",
    });
  });
  it("falls back to preferred_username then nickname", () => {
    expect(profileFromClaims({ preferred_username: "bob" }).display_name).toBe("bob");
    expect(profileFromClaims({ nickname: "bobby" }).display_name).toBe("bobby");
  });
  it("reads profile from a JWT payload", () => {
    const p = profileFromJwt(fakeJwt({ name: "Carol", picture: "https://img/c" }));
    expect(p?.display_name).toBe("Carol");
  });
});

describe("exchangeAuthCode / refreshWithIdp", () => {
  it("returns token on 200 and null on IdP error", async () => {
    const stub = stubFetch([
      idpTokenRoute(testEnv.OIDC_TOKEN_URL, { access_token: "at", expires_in: 3600 }),
    ]);
    restoreFetch = stub.restore;
    const tok = await exchangeAuthCode(testEnv, "code", "verifier", testEnv.OIDC_DESKTOP_REDIRECT_URI);
    expect(tok?.access_token).toBe("at");
    // client_secret must be sent to the IdP token endpoint (form body)
    const body = stub.calls[0].init?.body as URLSearchParams;
    expect(body.get("client_secret")).toBe(testEnv.OIDC_CLIENT_SECRET);
    expect(body.get("grant_type")).toBe("authorization_code");
  });

  it("refreshWithIdp returns null when IdP rejects", async () => {
    const stub = stubFetch([idpTokenRoute(testEnv.OIDC_TOKEN_URL, { error: "invalid_grant" }, 400)]);
    restoreFetch = stub.restore;
    expect(await refreshWithIdp(testEnv, "refresh")).toBeNull();
  });

  it("returns null when token response lacks access_token", async () => {
    const stub = stubFetch([idpTokenRoute(testEnv.OIDC_TOKEN_URL, { not_a_token: true })]);
    restoreFetch = stub.restore;
    expect(await refreshWithIdp(testEnv, "refresh")).toBeNull();
  });
});
